File: iOSKit/Core/UIApplication.m
//
// UIApplication.m
// ChangShou
//
// Created by Mudox on 01/03/2017.
// Copyright © 2017 Mudox. All rights reserved.
//

@import JacKit;

#import "The.h"

#ifdef DEBUG
static const DDLogLevel ddLogLevel = DDLogLevelVerbose;
#else
static const DDLogLevel ddLogLevel = DDLogLevelWarning;
#endif

#import "The.h"
#import "UIApplication.h"
#import "UIDevice.h"

@implementation UIApplication (MDX)

- (void)mdx_observeAppStates
{

  NSNotificationCenter *center = [NSNotificationCenter defaultCenter];
  NSDictionary         *dict   = @{
    // active state
    UIApplicationDidBecomeActiveNotification: @"did become active",
    UIApplicationWillResignActiveNotification: @"will resign active",
    // background state
    UIApplicationDidEnterBackgroundNotification: @"did enter background",
    UIApplicationWillEnterForegroundNotification: @"will enter foreground",

    // launching state
    UIApplicationDidFinishLaunchingNotification: @"did finish launching",

    // memory warning
    UIApplicationDidReceiveMemoryWarningNotification: @"did receive memory warning",

    // terminat state
    UIApplicationWillTerminateNotification: @"will terminate",

    // significant time changing
    UIApplicationSignificantTimeChangeNotification: @"siginificant time change",

    // status bar
    UIApplicationWillChangeStatusBarOrientationNotification: @"will change status bar orientation",
    UIApplicationDidChangeStatusBarOrientationNotification: @"did change status bar orientation",
    UIApplicationWillChangeStatusBarFrameNotification: @"will change status bar frame",
    UIApplicationDidChangeStatusBarFrameNotification: @"did change status bar frame",

    UIApplicationBackgroundRefreshStatusDidChangeNotification: @"did change background refresh status",

    // protected data
    UIApplicationProtectedDataWillBecomeUnavailable: @"prototected data will become unavailable",
    UIApplicationProtectedDataDidBecomeAvailable: @"protected data did become available",
  };

  [dict enumerateKeysAndObjectsUsingBlock:^(id _Nonnull name, id _Nonnull text, BOOL * _Nonnull stop) {
     [center addObserverForName:name object:[UIApplication sharedApplication] queue:nil usingBlock:^(NSNotification *_Nonnull note) {
        DDLogWarn(@"Application State] %@", text);
      }];
   }];
}

- (void)mdx_greet
{
  NSString *appName          = [theBundle objectForInfoDictionaryKey:(__bridge NSString *)kCFBundleNameKey];
  NSString *bundleIdentifier = theBundle.bundleIdentifier;
  NSString *version          = [theBundle objectForInfoDictionaryKey:@"CFBundleShortVersionString"];
  NSString *build            = [theBundle objectForInfoDictionaryKey:(__bridge NSString *)kCFBundleVersionKey];
  NSString *deviceName       = theDevice.name;
  NSString *deviceModel      = theDevice.localizedModel;
  NSString *isDebugMode      = BOOLSYMBOL(theApp.mdx_isInDebugMode);
  NSString *isSimulator      = BOOLSYMBOL(theDevice.mdx_isSimulator);
  NSString *systemName       = theDevice.systemName;
  NSString *systemVersion    = theDevice.systemVersion;

  NSString *lines = [
    @[
      [NSString stringWithFormat:@"App %@] launched", appName ],

      @">> [App]",
      [NSString stringWithFormat:@">>   bundle identifier: %@", bundleIdentifier],
      [NSString stringWithFormat:@">>     release version: %@", version],
      [NSString stringWithFormat:@">>       build version: %@", build],
      [NSString stringWithFormat:@">>       is debug mode: %@", isDebugMode],

      @">> [Device]",
      [NSString stringWithFormat:@">>         device name: %@", deviceName],
      [NSString stringWithFormat:@">>        device model: %@", deviceModel],
      [NSString stringWithFormat:@">>        is simulator: %@", isSimulator],

      @">> [System]",
      [NSString stringWithFormat:@">>         system name: %@", systemName],
      [NSString stringWithFormat:@">>      system version: %@", systemVersion],

      @"\n",
    ] componentsJoinedByString: @"\n"];

  DDLogInfo(@"%@", lines);
}

- (BOOL)mdx_isInDebugMode
{
#ifdef DEBUG
  return YES;
#else
  return NO;
#endif
}

@end

File: iOSKit/Core/UIApplication.swift
//
//  UIDevice.swift
//  CocoaLumberjack
//
//  Created by Mudox on 01/04/2018.
//

import UIKit

import JacKit
fileprivate let jack = Jack.with(levelOfThisFile: .verbose)

private let _appStates: [Notification.Name: String] = [
  // launching state
  Notification.Name.UIApplicationDidFinishLaunching: "App did finish launching",

  // active state
  Notification.Name.UIApplicationDidBecomeActive: "App did become active",
  Notification.Name.UIApplicationWillResignActive: "App will resign active",

  // background state
  Notification.Name.UIApplicationDidEnterBackground: "App did enter background",
  Notification.Name.UIApplicationWillEnterForeground: "App will enter foreground",

  // terminat state
  Notification.Name.UIApplicationWillTerminate: "App will terminate",

  // memory warning
  Notification.Name.UIApplicationDidReceiveMemoryWarning: "App did receive memory warning",

  // significant time changing
  Notification.Name.UIApplicationSignificantTimeChange: "App siginificant time change",

  // status bar
  Notification.Name.UIApplicationWillChangeStatusBarOrientation: "App will change status bar orientation",
  Notification.Name.UIApplicationDidChangeStatusBarOrientation: "App did change status bar orientation",
  Notification.Name.UIApplicationWillChangeStatusBarFrame: "App will change status bar frame",
  Notification.Name.UIApplicationDidChangeStatusBarFrame: "App did change status bar frame",

  // background refresh
  Notification.Name.UIApplicationBackgroundRefreshStatusDidChange: "App did change background refresh status",

  // protected data
  Notification.Name.UIApplicationProtectedDataWillBecomeUnavailable: "App prototected data will become unavailable",
  Notification.Name.UIApplicationProtectedDataDidBecomeAvailable: "App protected data did become available",
]


extension Mudoxive where Base: UIApplication {

  public var isRunningInDebugMode: Bool {
    #if DEBUG
      return true
    #else
      return false
    #endif
  }

  public var isRunningOnSimulator: Bool {
    #if targetEnvironment(simulator)
      return true
    #else
      return false
    #endif
  }

  public func greet() {
    fatalError("Unimplemented")
  }

  /**
   Log a line when App state transition happens.
   */
  public func observeAppStates() {
    _appStates.forEach { item in
      The.notificationCenter.addObserver(forName: item.key, object: nil, queue: nil, using: { _ in
        jack.debug(item.value, from: .custom("App state observer"))
      })
    }
  }

  public func dumpInfo() {
    let lines = """
      [App]
        - Name       :   \(The.info.appName)
        - ID         :   \(The.info.bundleID ?? "N/A")
        - Release    :   \(The.info.appRelease ?? "N/A")  (CFBundleShortVersionString)
        - Build      :   \(The.info.appBuild ?? "N/A") (kCFBundleVersionKey)
        - Debug      :   \(The.info.isDebug)
        - Simulator  :   \(The.info.isSimulator)
      [Device]
        - Name       :   \(The.info.deviceName)
        - Model      :   \(The.info.deviceModel)
        - UUID       :   \(The.info.deviceUUID ?? "N/A")
      [System]
        - Name       :   \(The.info.systemName)
        - Version    :   \(The.info.systemVersion)
      """
    jack.info(lines, from: .custom("App \(The.info.appName) Launched"))
  }
}
